src/daemon/link_connection.rs

```diff
@@ -36,41 +37,94 @@ use super::{
 #[derive(Clone)]
 pub struct LinkConnection {
-    mplex: Arc<Multiplex>,
     send_outgoing: Sender<RawPacket>,
     recv_incoming: Receiver<RawPacket>,
     remote_idpk: IdentityPublic,
-    _task: Arc<Immortal>,
+}
+
+pub struct LinkInfo {
+    pub client: Arc<LinkClient>,
+    pub connection_task: Immortal,
+    pub remote_pk: IdentityPublic,
 }
 
 impl LinkConnection {
     /// Creates a new Connection, from a single Pipe. Unlike in Geph, n2n Multiplexes in earendil all contain one pipe each.
-    pub async fn connect(ctx: DaemonContext, pipe: impl Pipe) -> anyhow::Result<Self> {
-        // First, we construct the Multiplex.
+    pub async fn connect(
+        ctx: DaemonContext,
+        pipe: impl Pipe,
+        their_fp: Option<Fingerprint>,
+    ) -> anyhow::Result<LinkInfo> {
         let my_mux_sk = MuxSecret::generate();
         let mplex = Arc::new(Multiplex::new(my_mux_sk, None));
         mplex.add_pipe(pipe);
         let (send_outgoing, recv_outgoing) = smol::channel::bounded(1);
         let (send_incoming, recv_incoming) = smol::channel::bounded(1);
-        let _task = Arc::new(Immortal::spawn(
-            connection_loop(ctx, mplex.clone(), send_incoming, recv_outgoing)
-                .unwrap_or_else(|e| panic!("connection_loop died with {:?}", e)),
-        ));
         let rpc = MultiplexRpcTransport::new(mplex.clone());
         let link = LinkClient::from(rpc);
+
+        let remote_pk_shared = Arc::new(Mutex::new(None));
+
+        let service = Arc::new(LinkService(LinkProtocolImpl {
+            ctx: ctx.clone(),
+            mplex: mplex.clone(),
+            remote_pk: remote_pk_shared.clone(),
+        }));
+
+        let task = Immortal::spawn(
+            connection_loop(mplex.clone(), send_incoming, recv_outgoing, service)
+                .unwrap_or_else(|e| panic!("connection_loop died with {:?}", e)),
+        );
+
         let resp = link
             .authenticate()
             .await
             .context("did not respond to authenticate")?;
+
         resp.verify(&mplex.peer_pk().context("could not obtain peer_pk")?)
             .context("did not authenticated correctly")?;
 
-        Ok(Self {
-            mplex,
+        let remote_fp = resp.full_pk.fingerprint();
+
+        let mut remote_pk = remote_pk_shared.lock();
+        *remote_pk = Some(resp.full_pk);
+
+        let conn = Self {
             send_outgoing,
             recv_incoming,
             remote_idpk: resp.full_pk,
-            _task,
+        };
+
+        if let Some(fp) = their_fp {
+            log::info!("about to insert into neightable for fp: {}", fp);
+
+            if fp != remote_fp {
+                anyhow::bail!(
+                    "out route fingerprint in config ({}), does not match link fingerprint: {}",
+                    fp,
+                    remote_fp
+                );
+            }
+
+            // add out_route entry into neightable
+            ctx.get(NEIGH_TABLE).insert_pinned(fp, conn);
```

Comment:
This is a very surprising side effect! I would not have imagined that constructing a LinkConnection would mutate something global like this.